hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/ratis/BaseRatisCommand.java

```diff
@@ -0,0 +1,132 @@
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_SECURITY_ENABLED_DEFAULT;
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_SECURITY_ENABLED_KEY;
+import static org.apache.ratis.shell.cli.RaftUtils.buildRaftGroupIdFromStr;
+import static org.apache.ratis.shell.cli.RaftUtils.buildRaftPeersFromStr;
+import static org.apache.ratis.shell.cli.RaftUtils.retrieveGroupInfoByGroupId;
+import static org.apache.ratis.shell.cli.RaftUtils.retrieveRemoteGroupId;
+
+/**
+ * The base class for all the ratis command.
+ */
+public class BaseRatisCommand {
+  private RaftGroup raftGroup;
+  private GroupInfoReply groupInfoReply;
+  private OzoneConfiguration config;
+  private String omServiceId;
+
+  public void run(String peers, String groupid, String omServiceId) throws Exception {
+    this.config = new OzoneConfiguration();
+    this.omServiceId = omServiceId;
+    List<RaftPeer> peerList = buildRaftPeersFromStr(peers);
+    RaftGroupId raftGroupIdFromConfig = buildRaftGroupIdFromStr(groupid);
+    raftGroup = RaftGroup.valueOf(raftGroupIdFromConfig, peerList);
+
+    try (RaftClient client = buildRaftClient(raftGroup)) {
+      RaftGroupId remoteGroupId = retrieveRemoteGroupId(raftGroupIdFromConfig, peerList, client, System.out);
+      groupInfoReply = retrieveGroupInfoByGroupId(remoteGroupId, peerList, client, System.out);
+      raftGroup = groupInfoReply.getGroup();
+    }
+
+  }
+
+  protected RaftClient buildRaftClient(RaftGroup raftGroup) throws Exception {
+    GrpcTlsConfig tlsConfig = null;
+    if (config.getBoolean(OZONE_SECURITY_ENABLED_KEY, OZONE_SECURITY_ENABLED_DEFAULT) &&
+        config.getBoolean(HDDS_GRPC_TLS_ENABLED, HDDS_GRPC_TLS_ENABLED_DEFAULT)) {
+      tlsConfig = createGrpcTlsConf(omServiceId);
+    }
+
+    return RatisHelper.newRaftClient(
+        SupportedRpcType.GRPC, null, raftGroup,
+        RatisHelper.createRetryPolicy(config), tlsConfig, config);
+  }
+
+  public RaftGroup getRaftGroup() {
+    return raftGroup;
+  }
+
+  public GroupInfoReply getGroupInfoReply() {
+    return groupInfoReply;
+  }
+
+  /**
+   * Get the leader id.
+   *
+   * @param roleInfo the role info
+   * @return the leader id
+   */
+  protected RaftPeerProto getLeader(RoleInfoProto roleInfo) {
+    if (roleInfo == null) {
+      return null;
+    }
+    if (roleInfo.getRole() == RaftPeerRole.LEADER) {
+      return roleInfo.getSelf();
+    }
+    FollowerInfoProto followerInfo = roleInfo.getFollowerInfo();
+    if (followerInfo == null) {
+      return null;
+    }
+    return followerInfo.getLeaderInfo().getId();
+  }
+
+  private GrpcTlsConfig createGrpcTlsConf(String omServiceId) throws Exception {
+    OzoneClient certClient = OzoneClientFactory.getRpcClient(omServiceId,
+        config);
+
+    ServiceInfoEx serviceInfoEx = certClient
+        .getObjectStore()
+        .getClientProxy()
+        .getOzoneManagerClient()
+        .getServiceInfo();
+
+    CACertificateProvider remoteCAProvider =
```

Comment:
I am not sure if this remoteProvider definition is intentional or is a misunderstanding of the ClientTrustManager's intents, but let me provide the intent based on which I hope you can decide, and course correct if needed.

The ClientTrustManager is designed in this way to ensure we can give the long running clients a possibility to renew their knowledge of the Ozone cluster's internal rootCA certificate that the client side uses to verify the identity of DataNodes in case of a rootCA rotation that happens some time before the rootCA actually expires, and is a long process by default during which there might be multiple rootCA certificates playing the role of the root of trust (the old one and the new one), after which period the new rootCA becomes the only valid one (but that does not affect the client side which can run further with the two together at the moment).

The current way of initializing the ClientTrustManager here, leads to a state where this renew does not happen, as both the remote and the in-memory provider is the same (the remote basically delegates to the in-memory instance via a lambda expression). This way of operation in the ClientTrustManager can be achieved by specifying the remote provider to null, as in both cases no additional RPC call would be done even if there is verification problem with the certificate (the new DN cert is signed by a different CA).

If this is a short running client which it is (as it runs one or more RPC calls in a short period of time and exits). If we can assume that these RPC calls are idempotent, then we don't need the remote provider, as in case of a rare event when two or more RPCs fall to the different sides on a rootCA rotation in time, then the whole thing can be done once more and it will not fail with a certificate issue the second time.

If this would be a long running client that issues multiple RPCs over a longer period of time, then it does worth it to provide a remote provider other than null, and in this case the best a remote provider can do to refresh the root of trust is to call OM's getServiceInfo method via an RPC, and provide the root(s) of trust based on the result of that RPC call.

Reply:
Thanks @fapifta  for the detailed explanation and recommendations regarding the ClientTrustManager and the remoteProvider configuration. In my case, I am indeed dealing with short-running clients that perform one RPC calls and then exit.
Based on your guidance, setting remoteProvider to null is an acceptable approach for my use case. But I'm open to any other suggestions or alternative approaches if other folks might have.

Reply:
I haven't calculated it exactly but if my memory serves, by default anything under 2 hours of runtime is safe without a remote provider, and in any reasonable setup problems will not arise with a client that does live for just a few seconds or even a few minutes, so yes, in this use case null should be completely fine.